[LowerToHw] Add symbol name to publicly visible sv::WireOp, and optimize wires with no symbols during Canonicalize

prithayan opened the pull request:
Add symbol name to wires during lowering from FIRRTL to HW. This is to ensure the publicly visible wires have a symbol attached to them. 
Changes in this PR:
- Attach symbols to non-temporary wires in lowerToHW
- Update the lit tests

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -887,6 +887,8 @@ struct FIRRTLLowering : public FIRRTLVisitor<FIRRTLLowering, LogicalResult> {
   // Create a temporary wire at the current insertion point, and try to
   // eliminate it later as part of lowering post processing.
   sv::WireOp createTmpWireOp(Type type, StringRef name) {
+    // This is a locally visible, private wire created by the compiler, so donot
```
typo "donot"

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -2204,6 +2211,8 @@ LogicalResult FIRRTLLowering::visitExpr(InvalidValueOp op) {
   // type.  We do that by lowering to a wire and return that.  As with the
   // SFC, we do not connect anything to this, because it is bidirectional.
   if (op.getType().isa<AnalogType>())
+    // This is a locally visible, private wire created by the compiler, so donot
```
typo donot

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -1582,6 +1587,8 @@ LogicalResult FIRRTLLowering::visitDecl(NodeOp op) {
   if (auto name = op->getAttrOfType<StringAttr>("name")) {
     if (!name.getValue().empty()) {
+      // This is a locally visible, private wire created by the compiler, so
+      // donot attach a symbol name.
```
same typo

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -1566,9 +1568,12 @@ LogicalResult FIRRTLLowering::visitDecl(WireOp op) {
+  // This is not a temporary wire created by the compiler, so attach a symbol
+  // name.
+  // TODO: How to ensure there's no symbol name conflict ?
+  return setLoweringTo<sv::WireOp>(op, resultType, nameAttr, nameAttr);
```
I don't think this is correct for the empty firrtl.wire name case.  This will pass a empty string attr that is non-null, which will be treated as a present symbol by the build.  I think you need to pass a `StringAttr()` not a `getStringAttr("")`

prithayan replied:
Thanks, fixed to use `StringAttr()` when name is empty.

prithayan commented:
Thanks @lattner  for the review, fixed the typos. 
Also updated the canonicalize to remove the temporary wires introduced by lowerToHW. And added the corresponding test cases.
Eventually we can remove the optimization logic from lowerToHW and only perform them in canonicalize.

darthscsi commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -1566,9 +1568,13 @@ LogicalResult FIRRTLLowering::visitDecl(WireOp op) {
+  auto symName = op.nameAttr() ? op.nameAttr() : StringAttr();
+  // This is not a temporary wire created by the compiler, so attach a symbol
+  // name.
+  // TODO: How to ensure there's no symbol name conflict ?
```
The existing name is unique in the module, but not in the symbol table.  I would do something like "__<module_name>_<wirename>".  Checking is a bit tricky due to SymbolTable not looking for symbol definitions in nested regions.

darthscsi commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -1743,8 +1751,16 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
+      auto wireName = ("." + portName + "." + field + ".wire").str();
+      // If no access to the subfield, then the wire cannot be optimized away,
+      // attach a symbol to it. It cannot be optimized because no rd/wr to it,
+      // no source value to replace it with.
```
Why?  synthetic wires with no read or write seem ideal to eliminate.

prithayan replied:
Was trying to handle the following case, we can't remove this wire. 
But I realized we should just handle this case during optimization, if wire read op has users, we cannot remove the wire. I will remove this symbol, and address your other comment in canonicalization.  
``` %.rw.wdata.wire = sv.wire  : !hw.inout<i42>
    %0 = sv.read_inout %.rw.wdata.wire : !hw.inout<i42>
    %_M.ro_data_0, %_M.rw_rdata_0 = hw.instance "_M" @FIRRTLMem_1_1_1_42_12_0_1_0(%clock1, %true, %c0_i4, %clock1, %true, %c0_i4, %true, %true, %0, %clock2, %inpred, %c0_i4, %true, %indata) : (i1, i1, i4, i1,
i1, i4, i1, i1, i42, i1, i1, i4, i1, i42) -> (i42, i42)```

darthscsi commented on lib/Dialect/SV/SVOps.cpp:
```diff
@@ -817,23 +817,58 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
-  rewriter.eraseOp(op);
+    // Remove all uses of the wire.
+    for (auto &use : make_early_inc_range(wire.getResult().getUses())) {
+      rewriter.eraseOp(use.getOwner());
```
What happens to the users of the read ops?  are they not invalid after this?

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -1566,9 +1568,13 @@ LogicalResult FIRRTLLowering::visitDecl(WireOp op) {
+  // Name attr is required on sv.wire but optional on firrtl.wire.
   auto nameAttr = op.nameAttr() ? op.nameAttr() : builder.getStringAttr("");
-  return setLoweringTo<sv::WireOp>(op, resultType, nameAttr);
+  auto symName = op.nameAttr() ? op.nameAttr() : StringAttr();
```
I think you just want `auto symName = op.nameAttr();` instead of the ternary.

lattner replied:
Eventually we should only add this symbol when there is a "dont touch" attribute present at the FIRRTL level.

prithayan replied:
> 
> 
> I think you just want `auto symName = op.nameAttr();` instead of the ternary.

Updated to prepend the module name to the symbol name, we need it to try and make the symbol name unique outside the module in the symbol table.

prithayan replied:
> 
> 
> Eventually we should only add this symbol when there is a "dont touch" attribute present at the FIRRTL level.

Sure, was planning create a separate PR to  handle the "dont touch" annotation

lattner commented on lib/Dialect/SV/SVOps.cpp:
```diff
@@ -817,23 +817,63 @@ void WireOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
-  rewriter.eraseOp(op);
+    bool canErase = true;
+    // Remove all uses of the wire.
+    for (auto &use : make_early_inc_range(wire.getResult().getUses())) {
```
Please split the canonicalize patch out to its own separate patch, as it is logically distinct and the other patch already has a big effect on the testsuite.

When this comes back in, please change it to just RAUW the "readinout" ops with an X value.  That way you can always remove the wire when there is no writes.

prithayan replied:
Thanks, created a separate PR for the canonicalize changes.

lattner commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -1566,9 +1568,24 @@ LogicalResult FIRRTLLowering::visitDecl(WireOp op) {
+                                      .getValue()
+                                      .str() +
+                                  "__" + nameAttr.getValue().str())
+          : StringAttr();
```
Please change his into an if statement to fix the crazy formatting:

```
 auto symName = op.nameAttr();
 if (symName) {
   auto moduleName = cast<hw::HWModuleOp>(op->getParentOp()).getName();
   symName = stuff;
}
```

prithayan replied:
Fixed the formatting.
